### ISSUE TITLE
Empty data.json

### ISSUE THREAD

**Issue by zorun** [wavexx/fgallery#48]
When running fgallery, it correctly loads and transform the images, but an empty `data.json` file is created.  Here is the output with 1.8.1:

```
$ fgallery photos output
reading completed        
processing completed        
generating archive...
hash- or arrayref expected (not a simple scalar, use allow_nonref to allow this) at /opt/fgallery/fgallery line 836.
```

This happens both with 14f3463 and tag 1.8.1.  The system is Archlinux, with perl 5.22.2 and perl-json 2.90.

**Comment by zorun** [comment]
The issue seems to be caused by `perl-json-xs`: when I desinstall this package, everything works as expected.  I think the JSON module looks for the json-xs implementation first and falls back to a built-in implementation.

I have no idea why the json-xs implementation fails in this case.  In a test script, using `encode_json` on a simple hash and printing the result works fine.

Possibly related: https://codedump.io/share/qmQOI7GQ1KN/1/strange-behavior-of-a-tied-hash-in-perl-when-asking-for-an-arrayref

**Comment by wavexx** [comment]
This looks like the same as #47, but at least I can dig this one more deeply. This is a nice find, but there must be something else too.

There are no tied hashes in the input to `encode_json`. Also, I do have JSON::XS installed (3.010 from debian unstable), and I cannot reproduce it here. Annoying. I'll try to install arch on a VM and see if I can have the same.

To @haennima: could you try to uninstall JSON::XS and see if this "fixes" the issue?

**Comment by zorun** [comment]
Indeed, it's the same issue as #47, sorry (the title wasn't really explicit).

The version of `perl-json-xs` in Archlinux is 3.02.

**Comment by wavexx** [comment]
On Sat, May 21 2016, zorun <notifications@github.com> wrote:
> Indeed, it's the same issue as #47, sorry (the title wasn't really explicit).
>
> The version of `perl-json-xs` in Archlinux is 3.02.

Ha, now *that* triggers the problem. In fact, it seems an issue with
JSON::XS 3.02 and threads.

The minimal test case:

  use JSON qw{encode_json};
  use threads;
  threads->create(sub {})->join();
  encode_json({})

I reported it upstream.

**Comment by wavexx** [comment]
After discussing with the author, is looks like JSON::XS does not support the current perl version and will probably never support threads.

I switched to Cpanel::JSON::XS (if available), or JSON::PP as a fallback and updated the instructions accordingly.

Could you check that it now works correctly?

**Comment by zorun** [comment]
Great, it works fine now, even when `perl-json-xs` is installed. Thanks!

Would you mind releasing a new minor version with this change?

**Comment by wavexx** [comment]
On Sun, May 22 2016, zorun <notifications@github.com> wrote:
> Great, it works fine now, even when `perl-json-xs` is installed. Thanks!

I basically had to, JSON::XS is required by several other modules :/

JSON::MaybeXS looked promising, but would actually also use JSON::XS if
available.

> Would you mind releasing a new minor version with this change?

I guess there's enough meat for 1.8.2.
Any particular reason you need it?

**Comment by zorun** [comment]
> I guess there's enough meat for 1.8.2.

Cool, thanks!

> Any particular reason you need it?

I am using the Archlinux package https://aur.archlinux.org/packages/fgallery/ and I don't like running programs from their git versions :)

I need perl-json-xs for `clive`, and there are probably other users of perl-json-xs: https://www.archlinux.org/packages/community/x86_64/perl-json-xs/

**Comment by wavexx** [comment]
> I am using the Archlinux package https://aur.archlinux.org/packages/fgallery/ and I don't like running programs from their git versions :)

Heh, I know the pain...

Let's give some hours to @haennima to check if this also works as expected.
I'll put 1.8.2 out tomorrow.

**Comment by wavexx** [comment]
Fixed in 1.8.2
